lib/jeos.pm

```diff
@@ -0,0 +1,60 @@
+# Copyright (C) 2015-2020 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+package jeos;
+use strict;
+use warnings;
+use parent 'Exporter';
+use testapi;
+use utils 'ensure_serialdev_permissions';
+use power_action_utils 'power_action';
+use Utils::Backends 'is_hyperv';
+use version_utils qw(is_sle is_tumbleweed is_leap);
+use bootloader_setup qw(change_grub_config grep_grub_settings grub_mkconfig set_framebuffer_resolution set_extrabootparams_grub_conf);
+
+our @EXPORT = qw(expect_mount_by_uuid set_grub_gfxmode rebooting is_efi_boot);
+
+sub expect_mount_by_uuid {
+    return (is_hyperv || is_sle('>=15-sp2') || is_tumbleweed || is_leap('>=15.2'));
+}
+
+# non empty */sys/firmware/efi/* must exist in UEFI mode
+sub is_efi_boot {
+    return !script_run('test -d /sys/firmware/efi/') && script_run('test -z "$(ls -A /sys/firmware/efi/)"');
+}
+
+sub rebooting {
+    my ($self, $msg) = @_;
+    power_action('reboot', textmode => 1);
+    record_info('reboot', $msg);
+    wait_serial '/^BdsDxe:\s+failed.*Access\sDenied/', expect_not_found => 1 or
```

Comment:
can we grep some log file for this instead using wait_serial ? because this would actually means that we freezing our test here for 90 seconds to make sure that all is fine .

Reply:
```
localhost login: YpuUt-0-
[   37.284607] reboot: Restarting system
[2J[01;01H[=3h[2J[01;01H[0m[35m[40m[0m[37m[40mBdsDxe: loading Boot0003 "opensuse" from HD(2,GPT,9DD7709A-086D-4B2B-A3B5-3B4451F0D520,0x1800,0x10800)/\EFI\opensuse\grubx64.efi
BdsDxe: failed to load Boot0003 "opensuse" from HD(2,GPT,9DD7709A-086D-4B2B-A3B5-3B4451F0D520,0x1800,0x10800)/\EFI\opensuse\grubx64.efi: Access Denied
BdsDxe: loading Boot0001 "UEFI Misc Device" from PciRoot(0x0)/Pci(0x8,0x0)
BdsDxe: starting Boot0001 "UEFI Misc Device" from PciRoot(0x0)/Pci(0x8,0x0)
UEFI SHIM
$Version: 14 $
$BuildMachine: GNU/Linux $
$Commit: 02e2fc61bd2fb7f0045f15db105de7b8ace3029f $
[0m[37m[44m[01;01H������������������������������������������������������������������������������Ŀ[02;01H�                            Verification succeeded   
```

I haven't found this anywhere. It is on serial line at very early boot stage before grub2 loads. ( I think )